packages/runtime-dom/__tests__/modules/class.spec.ts

```diff
@@ -0,0 +1,235 @@
+  }
+}
+
+describe('class', () => {
+  test('plain string', () => {
+    assertClass([
+      ['bar', 'foo bar'],
+      ['baz qux', 'foo baz qux'],
+      ['qux', 'foo qux'],
+      [undefined, 'foo']
+    ])
+  })
+
+  test('object value', () => {
+    assertClass([
+      [{ bar: true, baz: false }, 'foo bar'],
+      [{ baz: true }, 'foo baz'],
+      [null, 'foo'],
+      [{ 'bar baz': true, qux: false }, 'foo bar baz'],
+      [{ qux: true }, 'foo qux']
+    ])
+  })
+
+  test('array value', () => {
+    assertClass([
+      [['bar', 'baz'], 'foo bar baz'],
+      [['qux', 'baz'], 'foo qux baz'],
+      [['w', 'x y z'], 'foo w x y z'],
+      [undefined, 'foo'],
+      [['bar'], 'foo bar'],
+      [(val: Array<any>) => val.push('baz'), 'foo bar baz']
+    ])
+  })
+
+  test('array of mixed values', () => {
+    assertClass([
+      [['x', { y: true, z: true }], 'foo x y z'],
+      [['x', { y: true, z: false }], 'foo x y'],
+      [['f', { z: true }], 'foo f z'],
+      [['l', 'f', { n: true, z: true }], 'foo l f n z'],
+      [['x', {}], 'foo x'],
+      [undefined, 'foo']
+    ])
+  })
+
+  test('class merge between parent and child', () => {
+    const root = document.createElement('div')
+    type ClassItem = {
+      value: string | object | string[]
+    }
+    const childClass: ClassItem = { value: 'd' }
+
+    const child = () => h('div', { class: ['c', childClass.value] })
+
+    const parentClass: ClassItem = { value: 'b' }
+    const parent = () => h(child, { class: ['a', parentClass.value] })
+
+    render(h(parent), root)
+    expect(root.children[0].className).toBe('c a d b')
+
+    parentClass.value = 'e'
+    render(h(parent), root)
+    expect(root.children[0].className).toBe('c a d e')
+
+    parentClass.value = 'f'
+    render(h(parent), root)
+    expect(root.children[0].className).toBe('c a f e')
+
+    parentClass.value = { foo: true }
+    childClass.value = ['bar', 'baz']
+    expect(root.children[0].className).toBe('c a bar baz foo')
+  })
+
+  test('class merge between parent and child with props', () => {
+    const root = document.createElement('div')
+    type ClassItem = {
+      value: string | object | string[]
+    }
+    const childClass: ClassItem = { value: 'd' }
+
+    const child = {
+      setup() {
+        return () => h('div', { class: ['c', childClass.value] })
+      }
+    }
+
+    const parentClass: ClassItem = { value: 'b' }
+    const parent = {
+      setup() {
+        return () => h(child, { class: ['a', parentClass.value] })
+      }
+    }
+
+    const child2 = {
+      props: [],
+      setup() {
+        return () => h('div', { class: ['c', childClass.value] })
+      }
+    }
+    const parent2 = {
+      setup() {
+        return () => h(child2, { class: ['a', parentClass.value] })
+      }
+    }
+
+    render(h(parent2), root)
+    expect(root.children[0].className).toBe('c d a b')
+
+    render(h(parent), root)
+    expect(root.children[0].className).toBe('c d a b')
```

Comment:
The only difference between the 2 cases is that `child2` has an empty `props` declaration, therefore it successfully triggered class merging while the other one didn't.
I didn't see anything about this behavior in the `inheritAttrs` RFS.
Seems a bug in the current implementation?
@yyx990803